database2/ffldb/db.go

```diff
@@ -0,0 +1,2087 @@
+		return tx.pendingBlockData[idx].bytes, nil
+	}
+
+	// Lookup the location of the block in the files from the block index.
+	blockRow, err := tx.fetchBlockRow(hash)
+	if err != nil {
+		return nil, err
+	}
+	location := deserializeBlockLoc(blockRow)
+
+	// Read the block from the appropriate location.  The function also
+	// performs a checksum over the data to detect data corruption.
+	blockBytes, err := tx.db.store.readBlock(hash, location)
+	if err != nil {
+		return nil, err
+	}
+
+	return blockBytes, nil
+}
+
+// FetchBlocks returns the raw serialized bytes for the blocks identified by the
+// given hashes.  The raw bytes are in the format returned by Serialize on a
+// wire.MsgBlock.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockNotFound if any of the requested block hashed do not exist
+//   - ErrTxClosed if the transaction has already been closed
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// In addition, returns ErrDriverSpecific if any failures occur when reading the
+// block files.
+//
+// NOTE: The data returned by this function is only valid during a database
+// transaction.  Attempting to access it after a transaction has ended results
+// in undefined behavior.  This constraint prevents additional data copies and
+// allows support for memory-mapped database implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlocks(hashes []chainhash.Hash) ([][]byte, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	// NOTE: This could check for the existence of all blocks before loading
+	// any of them which would be faster in the failure case, however
+	// callers will not typically be calling this function with invalid
+	// values, so optimize for the common case.
+
+	// Load the blocks.
+	blocks := make([][]byte, len(hashes))
+	for i := range hashes {
+		var err error
+		blocks[i], err = tx.FetchBlock(&hashes[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return blocks, nil
+}
+
+// fetchPendingRegion attempts to fetch the provided region from any block which
+// are pending to be written on commit.  It will return nil for the byte slice
+// when the region references a block which is not pending.  When the region
+// does reference a pending block, it is bounds checked and returns
+// ErrBlockRegionInvalid if invalid.
+func (tx *transaction) fetchPendingRegion(region *database.BlockRegion) ([]byte, error) {
+	// Nothing to do if the block is not pending to be written on commit.
+	idx, exists := tx.pendingBlocks[*region.Hash]
+	if !exists {
+		return nil, nil
+	}
+
+	// Ensure the region is within the bounds of the block.
+	blockBytes := tx.pendingBlockData[idx].bytes
+	blockLen := uint32(len(blockBytes))
+	endOffset := region.Offset + region.Len
+	if endOffset < region.Offset || endOffset > blockLen {
```

Comment:
the check (endOffset < region.Offset) can be done much more cheaply at the beginning of the function, as it only requires the data contained in the passed region